### PR TITLE
Add missing model components and data misc

### PR DESCRIPTION
**Pull request by ayulockin**
- [x] Source/Target Embedding: In the paper same embedding is shared between both encoder and decoder. Since we are tokenising separately, maybe we create two different instance? will see.
- [x] Positional embedding: to be added to the token embedding
- [x] Decoder head (Linear + Softmax): compute token probability
- [ ] ~Label smoothing: regularisation applied during training - should be part of the linear head~ (add in the next PR for training)
- [x] residual dropout: a dropout add before adding x and residual
- [x] Shift output token right side

This PR also adds a Makefile to do lint/formatting.

### CODE REVIEW

**Comment by ariG23498** [comment]
Let me know when you would like for me to take a look at the PR

**Comment by ayulockin** [comment]
@ariG23498 ready for review.

File: CONTRIBUTION.md
Comment (ariG23498) [inline CONTRIBUTION.md:1]: I think this is very specific to us at the moment, and once we are open to collaborations and contributions we could create file and add in more details.

WDYT?

The current things in this file can rest in README (for us to use)

File: src/llm/wmt_data_utils.py
Comment (ariG23498) [inline src/llm/wmt_data_utils.py:144]: The comment is not clear to me.

All we are doing here is add the `<s>` special token to our target sequence.

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:19]: I think this is a redundant `nn.Dropout` Module.

It does not add the residual bit to the dropout bit.

I am open to keeping it, provided it does something along with applying dropout.

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:295]: Can you point to the place where we need to scale the embeddings? This is really interesting (and can form an ablation later)

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:317]: We should do away with the `print()` statements.

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:323]: I am not sure why we are doing this. Do we randomly dropout embeddings 👁️ ?

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:341]: Should we not include the newly built modules (embedding, lmhead, posembed) inside the Transformer Module?

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Left some comment. This is really good work.

File: CONTRIBUTION.md
Comment (ayulockin) [reply CONTRIBUTION.md:1]: yeah can move to the readme.

File: src/llm/wmt_data_utils.py
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:144]: because we are adding <s> we have shifted the output tokens by 1 unit position. do you want me to change the comment?

File: src/llm/wmt_model.py
Comment (ayulockin) [reply src/llm/wmt_model.py:19]: nice catch. moved residual (add) inside this class.

File: src/llm/wmt_model.py
Comment (ayulockin) [reply src/llm/wmt_model.py:291]: from the paper, 

"In the embedding layers, we multiply those weights by √dmodel."

image

File: src/llm/wmt_model.py
Comment (ayulockin) [reply src/llm/wmt_model.py:319]: I am doing because of my interpretation of this:

image

File: src/llm/wmt_data_utils.py
Comment (ariG23498) [reply src/llm/wmt_data_utils.py:144]: I think yes. It confuses me with the fact that shifting token by 1 is used for autoregressive models (decoder only)

File: src/llm/wmt_model.py
Comment (ariG23498) [reply src/llm/wmt_model.py:298]: Can we comment on top of the scaling bit "as suggested in the paper (arxiv link) in section 3.4"

File: src/llm/wmt_model.py
Comment (ayulockin) [reply src/llm/wmt_model.py:412]: After the change:

```
Transformer(
  (input_embedding): Embedding(
    (embedding): Embedding(1500, 64)
  )
  (output_embedding): Embedding(
    (embedding): Embedding(1500, 64)
  )
  (positional_encoding): PositionalEncoding(
    (residual_dropout): ResidualDropout(
      (dropout): Dropout(p=0.1, inplace=False)
    )
  )
  (encoder): Encoder(
    (encoder_layers): ModuleList(
      (0-1): 2 x EncoderBlock(
        (multi_head_attention): MultiHeadAttention(
          (W_out): Linear(in_features=64, out_features=64, bias=True)
        )
        (feed_forward): FeedForward(
          (layer1): Linear(in_features=64, out_features=256, bias=True)
          (activation): ReLU()
          (layer2): Linear(in_features=256, out_features=64, bias=True)
        )
        (layer_norm1): LayerNorm((64,), eps=1e-05, elementwise_affine=True)
        (layer_norm2): LayerNorm((64,), eps=1e-05, elementwise_affine=True)
        (residual_dropout): ResidualDropout(
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
  )
  (decoder): Decoder(
    (decoder_layers): ModuleList(
      (0-1): 2 x DecoderBlock(
        (masked_multi_head_attention): MultiHeadAttention(
          (W_out): Linear(in_features=64, out_features=64, bias=True)
        )
        (multi_head_attention): MultiHeadAttention(
          (W_out): Linear(in_features=64, out_features=64, bias=True)
        )
        (feed_forward): FeedForward(
          (layer1): Linear(in_features=64, out_features=256, bias=True)
          (activation): ReLU()
          (layer2): Linear(in_features=256, out_features=64, bias=True)
        )
        (layer_norm1): LayerNorm((64,), eps=1e-05, elementwise_affine=True)
        (layer_norm2): LayerNorm((64,), eps=1e-05, elementwise_affine=True)
        (layer_norm3): LayerNorm((64,), eps=1e-05, elementwise_affine=True)
        (residual_dropout): ResidualDropout(
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
  )
)
```

File: src/llm/wmt_model.py
Comment (ariG23498) [reply src/llm/wmt_model.py:326]: Very interesting. Thanks for pointing me to the source.

File: src/llm/wmt_model.py
Comment (ariG23498) [reply src/llm/wmt_model.py:260]: So you mean to say, each decoder block, gets as an input the same encoder output (that is the final encoder output)?

Comment (ayulockin) [reply src/llm/wmt_model.py:260]: yes.

**Comment by ayulockin** [comment]
@ariG23498 should we merge this if LGTM?

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:352]:
```suggestion
        self.source_embedding = Embedding(
```

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:355]:
```suggestion
        self.target_embedding = Embedding(
```

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:387]:
```suggestion
        encoder_input = self.source_embedding(encoder_input)
```

File: src/llm/wmt_model.py
Comment (ariG23498) [inline src/llm/wmt_model.py:391]:
```suggestion
        decoder_input = self.target_embedding(decoder_input)
```

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Small nits -- after you take care of them, let's merge it.

**Comment by ayulockin** [comment]
Ty!

**Pull request merged**
